File: .claude/merge-settings.py
#!/usr/bin/env python3
"""
Merges a Claude Code settings fragment into a target settings.json.
Idempotent — safe to run multiple times. Used by both the local-machine
install.sh and the GeoHavenOS SessionStart hook.

Usage: python3 merge-settings.py <fragment.json> <target-settings.json>
"""
import json, os, sys


def merge_stop_hooks(existing, fragment):
    existing_hooks = existing.setdefault("hooks", {})
    fragment_hooks = fragment.get("hooks", {})

    if "Stop" not in fragment_hooks:
        return

    existing_stop = existing_hooks.setdefault("Stop", [])
    for fragment_entry in fragment_hooks["Stop"]:
        matcher = fragment_entry.get("matcher", "")
        for fragment_hook in fragment_entry.get("hooks", []):
            cmd = fragment_hook.get("command", "")
            # Skip if this exact command is already present anywhere in Stop hooks
            already_present = any(
                h.get("command") == cmd
                for entry in existing_stop
                for h in entry.get("hooks", [])
            )
            if already_present:
                continue
            # Append to existing entry with same matcher, or create a new entry
            found = False
            for entry in existing_stop:
                if entry.get("matcher", "") == matcher:
                    entry.setdefault("hooks", []).append(fragment_hook)
                    found = True
                    break
            if not found:
                existing_stop.append({"matcher": matcher, "hooks": [fragment_hook]})


def merge_permissions(existing, fragment):
    if "permissions" not in fragment:
        return
    existing_perms = existing.setdefault("permissions", {})
    for key, values in fragment["permissions"].items():
        if isinstance(values, list):
            existing_list = existing_perms.setdefault(key, [])
            for item in values:
                if item not in existing_list:
                    existing_list.append(item)


def main():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <fragment.json> <target-settings.json>")
        sys.exit(1)

    fragment_path = sys.argv[1]
    settings_path = sys.argv[2]

    with open(fragment_path) as f:
        fragment = json.load(f)

    try:
        with open(settings_path) as f:
            existing = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        existing = {}

    merge_stop_hooks(existing, fragment)
    merge_permissions(existing, fragment)

    os.makedirs(os.path.dirname(os.path.abspath(settings_path)), exist_ok=True)
    with open(settings_path, "w") as f:
        json.dump(existing, f, indent=4)
        f.write("\n")

    print(f"Merged into {settings_path}")


if __name__ == "__main__":
    main()

File: .claude/merge-settings.js
#!/usr/bin/env node
/**
 * merge-settings.js — Node.js equivalent of merge-settings.py
 * Merges a Claude Code settings fragment into a target settings.json.
 * Idempotent — safe to run multiple times.
 *
 * Usage: node merge-settings.js <fragment.json> <target-settings.json>
 */
const fs = require('fs');
const path = require('path');

const [, , fragmentPath, settingsPath] = process.argv;
if (!fragmentPath || !settingsPath) {
  console.error(`Usage: node merge-settings.js <fragment.json> <target-settings.json>`);
  process.exit(1);
}

const fragment = JSON.parse(fs.readFileSync(fragmentPath, 'utf8'));

let existing = {};
try {
  existing = JSON.parse(fs.readFileSync(settingsPath, 'utf8'));
} catch (_) {
  // File missing or invalid JSON — start fresh
}

// Merge Stop hooks
if (fragment.hooks?.Stop) {
  if (!existing.hooks) existing.hooks = {};
  if (!existing.hooks.Stop) existing.hooks.Stop = [];
  const existingStop = existing.hooks.Stop;

  for (const fragEntry of fragment.hooks.Stop) {
    const matcher = fragEntry.matcher ?? '';
    for (const hook of fragEntry.hooks ?? []) {
      const cmd = hook.command;
      const alreadyPresent = existingStop.some(e =>
        (e.hooks ?? []).some(h => h.command === cmd)
      );
      if (alreadyPresent) continue;

      const target = existingStop.find(e => (e.matcher ?? '') === matcher);
      if (target) {
        target.hooks = target.hooks ?? [];
        target.hooks.push(hook);
      } else {
        existingStop.push({ matcher, hooks: [hook] });
      }
    }
  }
}

// Merge permissions.allow
if (fragment.permissions?.allow) {
  if (!existing.permissions) existing.permissions = {};
  if (!existing.permissions.allow) existing.permissions.allow = [];
  for (const item of fragment.permissions.allow) {
    if (!existing.permissions.allow.includes(item)) {
      existing.permissions.allow.push(item);
    }
  }
}

fs.mkdirSync(path.dirname(path.resolve(settingsPath)), { recursive: true });
fs.writeFileSync(settingsPath, JSON.stringify(existing, null, 4) + '\n');
console.log(`Merged into ${settingsPath}`);
